go/consensus/cometbft/apps/keymanager/churp/txs.go

```diff
@@ -0,0 +1,253 @@
+		Committee:       nil,
+		Applications:    nil,
+		Checksum:        nil,
+	}
+
+	if err := state.SetStatus(ctx, &status); err != nil {
+		ctx.Logger().Error("keymanager: churp: failed to set status",
+			"err", err,
+		)
+		return fmt.Errorf("keymanager: churp: failed to set status: %w", err)
+	}
+
+	ctx.EmitEvent(tmapi.NewEventBuilder(ext.appName).TypedAttribute(&churp.CreateEvent{
+		Status: &status,
+	}))
+
+	return nil
+}
+
+func (ext *churpExt) update(ctx *tmapi.Context, req *churp.UpdateRequest) error {
+	// Prepare state.
+	state := churpState.NewMutableState(ctx.State())
+
+	// Ensure that the runtime exists and is a key manager.
+	kmRt, err := common.KeyManagerRuntime(ctx, req.RuntimeID)
+	if err != nil {
+		return err
+	}
+
+	// Ensure that the tx signer is the key manager owner.
+	if !kmRt.EntityID.Equal(ctx.TxSigner()) {
+		return fmt.Errorf("keymanager: churp: invalid signer")
+	}
+
+	// Get the existing status.
+	status, err := state.Status(ctx, req.RuntimeID, req.ID)
+	if err != nil {
+		return fmt.Errorf("keymanager: churp: non-existing ID: %d", req.ID)
+	}
+
+	// Verify data.
+	if err = req.ValidateBasic(); err != nil {
+		return fmt.Errorf("keymanager: churp: invalid config: %w", err)
+	}
+
+	// Handle handoff interval change.
+	if req.HandoffInterval != nil {
+		switch {
+		case *req.HandoffInterval == 0:
+			// Cancel and disable handoffs.
+			status.NextHandoff = churp.HandoffsDisabled
+			status.Applications = nil
+			status.Checksum = nil
+		case status.HandoffInterval == 0:
+			// Schedule the next handoff.
+			status.NextHandoff, err = ext.computeNextHandoff(ctx)
+			if err != nil {
+				return err
+			}
+		default:
+			// Preserve the scheduled handoff and transition
+			// to the new interval thereafter.
+		}
+		status.HandoffInterval = *req.HandoffInterval
+	}
+
+	// Handle policy change.
+	if req.Policy != nil {
+		if err = req.Policy.SanityCheck(&status.Policy.Policy); err != nil {
+			return fmt.Errorf("keymanager: churp: invalid config: %w", err)
+		}
+		status.Policy = *req.Policy
+	}
+
+	if ctx.IsCheckOnly() {
+		return nil
+	}
+
+	// Charge gas for this operation.
+	kmParams, err := state.ConsensusParameters(ctx)
+	if err != nil {
+		return err
+	}
+	if err = ctx.Gas().UseGas(1, churp.GasOpUpdate, kmParams.GasCosts); err != nil {
+		return err
+	}
+
+	// Return early if simulating since this is just estimating gas.
+	if ctx.IsSimulation() {
+		return nil
+	}
+
+	if err := state.SetStatus(ctx, status); err != nil {
+		ctx.Logger().Error("keymanager: churp: failed to set status",
+			"err", err,
+		)
+		return fmt.Errorf("keymanager: churp: failed to set status: %w", err)
+	}
+
+	ctx.EmitEvent(tmapi.NewEventBuilder(ext.appName).TypedAttribute(&churp.UpdateEvent{
+		Status: status,
+	}))
+
+	return nil
+}
+
+func (ext *churpExt) onEpochChange(ctx *tmapi.Context, epoch beacon.EpochTime) error {
+	state := churpState.NewMutableState(ctx.State())
+
+	statuses, err := state.Statuses(ctx)
```

Comment:
This needs to somehow at least skip suspended runtimes, otherwise this list can only grow (even if additional stake requirements are not met and runtime is suspended) and iteration will be slower and slower?